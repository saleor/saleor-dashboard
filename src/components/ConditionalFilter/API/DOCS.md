# API Providers & Handlers

This document explains how to provide data for filter options using API Providers and Handlers.

## Overview

When a user clicks on a filter, they expect a list of options to choose from (e.g., a list of channels, categories, or simple "Yes/No" options).

**Handlers** are responsible for fetching this data (or simply returning a pre-defined list of options).

**API Provider** is a domain-specific component that determines which `Handler` to use for a given filter field.

## Handlers

Handlers are classes that implement a simple `fetch` method, which returns a promise resolving to an array of `ItemOption`.

| Handler Class                                               | Fetches...                       | Use Case                                      |
| ----------------------------------------------------------- | -------------------------------- | --------------------------------------------- |
| `NoopValuesHandler`                                         | Nothing. Returns an empty array. | Free-text fields like Order ID, search terms. |
| `BooleanValuesHandler`                                      | Static "Yes" and "No" options.   | `isActive`, `isPublished`, etc.               |
| `EnumValuesHandler`                                         | Options from a GraphQL enum.     | `OrderStatus`, `PaymentChargeStatusEnum`.     |
| Custom Handlers (e.g., `ChannelHandler`, `CategoryHandler`) | Dynamic data from the API.       | Lists of channels, categories, products.      |

## How to Add a New Field's Data Provider

### Step 1: Create a Handler (if needed)

If your field requires fetching dynamic data from the API, create a new handler.

```typescript
// src/components/ConditionalFilter/API/handlers/MyEntityHandler.ts

export class MyEntityHandler implements Handler {
  constructor(
    private client: ApolloClient<unknown>,
    private query: string,
  ) {}

  async fetch(): Promise<ItemOption[]> {
    const { data } = await this.client.query<MyQuery, MyQueryVariables>({
      query: MY_ENTITY_QUERY,
      variables: {
        first: 10, // Always paginate
        filter: { search: this.query },
      },
    });

    return data.myEntities.edges.map(({ node }) => ({
      label: node.name,
      value: node.id,
      slug: node.slug,
    }));
  }
}
```

### Step 2: Update the API Provider

Each domain (e.g., Products, Orders) has its own API Provider hook. Update the relevant provider to use your new handler for the corresponding field.

```typescript
// src/components/ConditionalFilter/API/providers/OrderFilterAPIProvider.tsx

function createAPIHandler(
  element: FilterElement,
  inputValue: string,
  client: ApolloClient<unknown>,
): Handler {
  const rowType = element.rowType();

  switch (rowType) {
    // ... existing cases ...

    case "myNewField":
      return new MyEntityHandler(client, inputValue);

    default:
      // Fallback for fields that don't need options
      return new NoopValuesHandler([]);
  }
}
```

Example in orders: `useOrderFilterAPIProvider` uses this `createAPIHandler` function to dispatch to the correct handler.

## Key Patterns & Pitfalls

1.  **Error Handling**: Handlers should catch any errors from API calls and return an empty array `[]` to prevent the UI from crashing. Use optional chaining (`?.`) when accessing nested data from a response.
2.  **Performance**:
    - **Debounce Input**: For handlers that take search input, the input should be debounced in the UI before being passed to the handler to avoid excessive API calls while the user is typing.
    - **Limiting**: Remember to fetch only a few options from API using `limit:` parameter.
3.  **Using resolvers** use a resolver pattern where the `QueryVarsBuilder` for a field is also responsible for creating the `Handler`. This keeps related logic together.
    - Note some olders views might use a simpler `switch` statement in the API Provider to select the `Handler`. This is easier for simpler cases. Choose the pattern that fits the complexity of your domain.
