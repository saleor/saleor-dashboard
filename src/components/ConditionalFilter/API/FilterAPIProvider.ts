import { FilterContainer } from "../FilterElement";
import { ItemOption } from "../FilterElement/ConditionValue";
import { LeftOperand } from "../LeftOperandsProvider";

export interface FilterAPIProvider {
  fetchRightOptions: (
    position: string,
    value: FilterContainer,
    inputValue: string,
  ) => Promise<ItemOption[]>;
  fetchLeftOptions: (inputValue: string) => Promise<LeftOperand[]>;
}
