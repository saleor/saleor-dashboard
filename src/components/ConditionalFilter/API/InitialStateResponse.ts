import { AttributeInputTypeEnum } from "@dashboard/graphql";

import { createBooleanOption } from "../constants";
import { AttributeInputType } from "../FilterElement/ConditionOptions";
import { ItemOption } from "../FilterElement/ConditionValue";
import { UrlToken } from "../ValueProvider/UrlToken";

export interface AttributeDTO {
  choices: Array<{
    label: string;
    value: string;
    slug: string;
    originalSlug?: string;
  }>;
  inputType: AttributeInputType;
  label: string;
  slug: string;
  value: string;
}

export interface InitialState {
  category: ItemOption[];
  attribute: Record<string, AttributeDTO>;
  channel: ItemOption[];
  collection: ItemOption[];
  productType: ItemOption[];
  isAvailable: ItemOption[];
  isPublished: ItemOption[];
  isVisibleInListing: ItemOption[];
  hasCategory: ItemOption[];
  giftCard: ItemOption[];
}

export class InitialStateResponse implements InitialState {
  constructor(
    public category: ItemOption[] = [],
    public attribute: Record<string, AttributeDTO> = {},
    public channel: ItemOption[] = [],
    public collection: ItemOption[] = [],
    public productType: ItemOption[] = [],
    public isAvailable: ItemOption[] = [],
    public isPublished: ItemOption[] = [],
    public isVisibleInListing: ItemOption[] = [],
    public hasCategory: ItemOption[] = [],
    public giftCard: ItemOption[] = [],
  ) {}

  public attributeByName(name: string) {
    return this.attribute[name];
  }

  public static empty() {
    return new InitialStateResponse();
  }

  public filterByUrlToken(token: UrlToken) {
    if (token.isAttribute() && token.hasDynamicValues()) {
      return this.attribute[token.name].choices.filter(({ value }) =>
        token.value.includes(value),
      );
    }

    if (token.isAttribute()) {
      const attr = this.attribute[token.name];
      return attr.inputType === "BOOLEAN"
        ? createBooleanOption(
            token.value === "true",
            AttributeInputTypeEnum.BOOLEAN,
          )
        : token.value;
    }

    if (!token.isLoadable()) {
      return [token.value] as string[];
    }

    return this.getEntryByName(token.name).filter(
      ({ slug }) => slug && token.value.includes(slug),
    );
  }

  private getEntryByName(name: string) {
    switch (name) {
      case "category":
        return this.category;
      case "collection":
        return this.collection;
      case "productType":
        return this.productType;
      case "channel":
        return this.channel;
      case "isAvailable":
        return this.isAvailable;
      case "isPublished":
        return this.isPublished;
      case "isVisibleInListing":
        return this.isVisibleInListing;
      case "hasCategory":
        return this.hasCategory;
      case "giftCard":
        return this.giftCard;
      default:
        return [];
    }
  }
}
