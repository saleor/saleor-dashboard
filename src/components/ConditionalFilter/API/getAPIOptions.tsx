// @ts-strict-ignore
import { ApolloClient } from "@apollo/client";
import {
  _GetAttributeChoicesDocument,
  _GetCategoriesChoicesDocument,
  _GetChannelOperandsDocument,
  _GetCollectionsChoicesDocument,
  _GetDynamicLeftOperandsDocument,
  _GetProductTypesChoicesDocument,
} from "@dashboard/graphql";

import { FilterElement } from "../FilterElement";
import { FilterContainer } from "../useFilterContainer";

const getFilterElement = (value: any, index: number): FilterElement => {
  const possibleFilterElement = value[index];
  return typeof possibleFilterElement != "string"
    ? possibleFilterElement
    : null;
};

export const getInitialRightOperatorOptions = async (
  client: ApolloClient<any>,
  position: string,
  value: FilterContainer,
) => {
  const index = parseInt(position, 10);
  const filterElement = getFilterElement(value, index);

  if (filterElement.isAttribute()) {
    const { data } = await client.query({
      query: _GetAttributeChoicesDocument,
      variables: {
        slug: filterElement.value.value,
        first: 5,
        query: "",
      },
    });
    return data.attribute.choices.edges.map(({ node }) => ({
      label: node.name,
      value: node.id,
      slug: node.slug,
    }));
  }

  if (filterElement.isCollection()) {
    const { data } = await client.query({
      query: _GetCollectionsChoicesDocument,
      variables: {
        first: 5,
        query: "",
      },
    });

    return data.collections.edges.map(({ node }) => ({
      label: node.name,
      value: node.id,
      slug: node.slug,
    }));
  }

  if (filterElement.isCategory()) {
    const { data } = await client.query({
      query: _GetCategoriesChoicesDocument,
      variables: {
        first: 5,
        query: "",
      },
    });

    return data.categories.edges.map(({ node }) => ({
      label: node.name,
      value: node.id,
      slug: node.slug,
    }));
  }

  if (filterElement.isProductType()) {
    const { data } = await client.query({
      query: _GetProductTypesChoicesDocument,
      variables: {
        first: 5,
        query: "",
      },
    });

    return data.productTypes.edges.map(({ node }) => ({
      label: node.name,
      value: node.id,
      slug: node.slug,
    }));
  }

  if (filterElement.isChannel()) {
    const { data } = await client.query({
      query: _GetChannelOperandsDocument,
    });

    return data.channels.map(({ id, name, slug }) => ({
      label: name,
      value: id,
      slug,
    }));
  }
};

export const getRightOperatorOptionsByQuery = async (
  client: ApolloClient<any>,
  position: string,
  value: FilterContainer,
  inputValue: string,
) => {
  const index = parseInt(position, 10);
  const filterElement = getFilterElement(value, index);

  if (filterElement.isAttribute()) {
    const { data } = await client.query({
      query: _GetAttributeChoicesDocument,
      variables: {
        slug: filterElement.value.value,
        first: 5,
        query: inputValue,
      },
    });
    return data.attribute.choices.edges.map(({ node }) => ({
      label: node.name,
      value: node.id,
    }));
  }

  if (filterElement.isCollection()) {
    const { data } = await client.query({
      query: _GetCollectionsChoicesDocument,
      variables: {
        first: 5,
        query: inputValue,
      },
    });

    return data.collections.edges.map(({ node }) => ({
      label: node.name,
      value: node.id,
      slug: node.slug,
    }));
  }

  if (filterElement.isCategory()) {
    const { data } = await client.query({
      query: _GetCategoriesChoicesDocument,
      variables: {
        first: 5,
        query: inputValue,
      },
    });

    return data.categories.edges.map(({ node }) => ({
      label: node.name,
      value: node.id,
      slug: node.slug,
    }));
  }

  if (filterElement.isProductType()) {
    const { data } = await client.query({
      query: _GetProductTypesChoicesDocument,
      variables: {
        first: 5,
        query: inputValue,
      },
    });

    return data.productTypes.edges.map(({ node }) => ({
      label: node.name,
      value: node.id,
    }));
  }

  if (filterElement.isChannel()) {
    const { data } = await client.query({
      query: _GetChannelOperandsDocument,
    });
    const options = data.channels.map(({ id, name, slug }) => ({
      label: name,
      value: id,
      slug,
    }));

    return options.filter(({ label }) =>
      label.toLowerCase().includes(inputValue.toLowerCase()),
    );
  }
};

export const getLeftOperatorOptions = async (
  client: any,
  inputValue: string,
) => {
  const { data } = await client.query({
    query: _GetDynamicLeftOperandsDocument,
    variables: {
      first: 5,
      query: inputValue,
    },
  });
  return data.attributes.edges.map(({ node }) => ({
    label: node.name,
    value: node.id,
    type: node.inputType,
    slug: node.slug,
  }));
};
