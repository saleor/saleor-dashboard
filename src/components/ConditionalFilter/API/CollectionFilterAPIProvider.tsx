import { ApolloClient, useApolloClient } from "@apollo/client";

import { RowType } from "../constants";
import { FilterContainer, FilterElement } from "../FilterElement";
import { FilterAPIProvider } from "./FilterAPIProvider";
import { BooleanValuesHandler, ChannelHandler, Handler, NoopValuesHandler } from "./Handler";

const isStaticBoolean = (rowType: RowType) => {
  return ["published"].includes(rowType);
};

const getFilterElement = (value: FilterContainer, index: number): FilterElement => {
  const possibleFilterElement = value[index];

  if (typeof possibleFilterElement !== "string" && !Array.isArray(possibleFilterElement)) {
    return possibleFilterElement;
  }

  throw new Error("Unknown filter element used to create API handler");
};
const createAPIHandler = (
  selectedRow: FilterElement,
  client: ApolloClient<unknown>,
  inputValue: string,
): Handler => {
  const rowType = selectedRow.rowType();

  if (rowType && isStaticBoolean(rowType) && rowType !== "attribute") {
    return new BooleanValuesHandler([
      {
        label: "True",
        value: "PUBLISHED",
        type: rowType,
        slug: "published",
      },
      {
        label: "False",
        value: "HIDDEN",
        type: rowType,
        slug: "hidden",
      },
    ]);
  }

  if (rowType === "ids") {
    return new NoopValuesHandler([]);
  }

  if (rowType === "metadata") {
    return new NoopValuesHandler([]);
  }

  if (rowType === "channel") {
    return new ChannelHandler(client, inputValue);
  }

  throw new Error(`Unknown filter element: "${rowType}"`);
};

export const useCollectionFilterAPIProvider = (): FilterAPIProvider => {
  const client = useApolloClient();

  const fetchRightOptions = async (
    position: string,
    value: FilterContainer,
    inputValue: string,
  ) => {
    const index = parseInt(position, 10);
    const filterElement = getFilterElement(value, index);

    const handler = createAPIHandler(filterElement, client, inputValue);

    return handler.fetch();
  };
  const fetchLeftOptions = async () => {
    return [];
  };

  return {
    fetchRightOptions,
    fetchLeftOptions,
  };
};
