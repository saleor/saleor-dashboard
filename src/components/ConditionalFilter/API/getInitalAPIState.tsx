// @ts-strict-ignore
import { useApolloClient } from "@apollo/client";
import {
  _GetChannelOperandsDocument,
  _SearchAttributeOperandsDocument,
  _SearchCategoriesOperandsDocument,
  _SearchCollectionsOperandsDocument,
  _SearchProductTypesOperandsDocument,
} from "@dashboard/graphql";
import { useEffect, useState } from "react";

import { InitialStateResponse } from "./InitialStateResponse";

interface Props {
  category?: string[];
  collection?: string[];
  channel?: string[];
  producttype?: string[];
  attribute?: {
    [attribute: string]: string[];
  };
}

export const useInitialAPIState = ({
  category = [],
  collection = [],
  producttype = [],
  channel = [],
  attribute = {},
}: Props) => {
  const client = useApolloClient();
  const [data, setData] = useState([]);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    const queriesToRun = [];

    const fetchQueries = async () => {
      const data = await Promise.all(queriesToRun);
      setData(data);
      setLoading(false);
    };

    if (channel.length > 0) {
      queriesToRun.push(
        client.query({
          query: _GetChannelOperandsDocument,
        }),
      );
    } else {
      queriesToRun.push({});
    }

    if (collection.length > 0) {
      queriesToRun.push(
        client.query({
          query: _SearchCollectionsOperandsDocument,
          variables: {
            collectionsSlugs: collection,
            first: collection.length,
          },
        }),
      );
    } else {
      queriesToRun.push({});
    }

    if (category.length > 0) {
      queriesToRun.push(
        client.query({
          query: _SearchCategoriesOperandsDocument,
          variables: {
            categoriesSlugs: category,
            first: category.length,
          },
        }),
      );
    } else {
      queriesToRun.push({});
    }

    if (producttype.length > 0) {
      queriesToRun.push(
        client.query({
          query: _SearchProductTypesOperandsDocument,
          variables: {
            productTypesSlugs: producttype,
            first: producttype.length,
          },
        }),
      );
    } else {
      queriesToRun.push({});
    }

    if (Object.keys(attribute).length > 0) {
      queriesToRun.push(
        client.query({
          query: _SearchAttributeOperandsDocument,
          variables: {
            attributesSlugs: Object.keys(attribute),
            choicesIds: Object.values(attribute).flat(),
            first: Object.keys(attribute).length,
          },
        }),
      );
    } else {
      queriesToRun.push({});
    }

    fetchQueries();
  }, []);

  const [
    channelData,
    collectionData,
    categoryData,
    productTypesData,
    attributesData,
  ] = data;

  const channelPicks =
    channelData?.data?.channels
      ?.filter(({ slug }) => channel.includes(slug))
      .map(({ id, name }) => ({ label: name, value: id })) ?? [];

  const collectionPicks =
    collectionData?.data?.search?.edges.map(({ node }) => ({
      label: node?.name,
      value: node?.id,
      slug: node?.slug,
    })) ?? [];

  const categoryPicks =
    categoryData?.data?.search?.edges.map(({ node }) => ({
      label: node?.name,
      value: node?.id,
      slug: node?.slug,
    })) ?? [];

  const productTypePicks =
    productTypesData?.data?.search?.edges.map(({ node }) => ({
      label: node?.name,
      value: node?.id,
      slug: node?.slug,
    })) ?? [];

  const attributePicks =
    attributesData?.data?.search?.edges.reduce(
      (acc, { node }) => ({
        ...acc,
        [node?.slug]: {
          choices: node?.choices.edges.map(({ node }) => ({
            label: node?.name,
            value: node?.id,
            slug: node?.slug,
          })),
          slug: node?.slug,
          value: node?.id,
          label: node?.name,
          inputType: node?.inputType,
        },
      }),
      {},
    ) ?? {};

  return {
    data: new InitialStateResponse(
      categoryPicks,
      attributePicks,
      channelPicks,
      collectionPicks,
      productTypePicks,
    ),
    loading,
  };
};
