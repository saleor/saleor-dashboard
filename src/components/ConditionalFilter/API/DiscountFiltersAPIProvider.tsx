import { FilterAPIProvider } from "@dashboard/components/ConditionalFilter/API/FilterAPIProvider";

export const useDiscountFilterAPIProvider = (): FilterAPIProvider => {
  const fetchRightOptions = async () => {
    return [];
  };

  const fetchLeftOptions = async () => {
    return [];
  };

  return {
    fetchRightOptions,
    fetchLeftOptions,
  };
};
