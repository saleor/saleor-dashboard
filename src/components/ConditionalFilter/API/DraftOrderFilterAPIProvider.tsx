import { FilterAPIProvider } from "@dashboard/components/ConditionalFilter/API/FilterAPIProvider";

export const useDraftOrderFilterAPIProvider = (): FilterAPIProvider => {
  const fetchRightOptions = async () => {
    return [];
  };
  const fetchLeftOptions = async () => {
    return [];
  };

  return {
    fetchRightOptions,
    fetchLeftOptions,
  };
};
