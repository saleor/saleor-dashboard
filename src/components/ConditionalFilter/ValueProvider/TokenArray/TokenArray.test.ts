import { InitialStateResponse } from "../../API/InitialStateResponse";
import { TokenArray } from ".";

describe("ConditionalFilter / ValueProvider / TokenArray", () => {
  it("should parse empty params", () => {
    // Arrange
    const url = new TokenArray("");
    // Act
    const fetchingParams = url.getFetchingParams();
    // Assert
    expect(fetchingParams).toEqual({
      category: [],
      collection: [],
      channel: [],
      productType: [],
      attribute: {},
    });
  });

  it("should parse params with values", () => {
    // Arrange
    const params = new URLSearchParams({
      "0[s0.price]": "123",
      "1": "AND",
      "2[s0.channel]": "channel-pln",
      "3": "OR",
      "4[s2.collection][0]": "featured-products",
      "5": "AND",
      "6[s0.productType]": "beer",
      "7": "AND",
      "8[s2.category][0]": "accessories",
      "9[s2.category][1]": "groceries",
      "10": "AND",
      "11[o2.bottle-size][0]": "attribute-id",
    });
    // Act
    const url = new TokenArray(params.toString());
    const fetchingParams = url.getFetchingParams();
    // Assert
    expect(fetchingParams).toEqual({
      attribute: {
        "bottle-size": ["attribute-id"],
      },
      category: ["accessories", "groceries"],
      channel: ["channel-pln"],
      collection: ["featured-products"],
      productType: ["beer"],
    });
  });

  it("should create filter container from a given response", () => {
    // Arrange
    const params = new URLSearchParams({
      "0[s0.price]": "123",
      "1": "AND",
      "2[s0.channel]": "channel-pln",
    });
    const response = new InitialStateResponse(
      [
        {
          label: "Cat1",
          value: "cat-1-id",
          slug: "cat1",
        },
      ],
      {},
      [
        {
          label: "Channel PLN",
          value: "channel-pln",
          slug: "channel-pln",
        },
      ],
    );

    // Act
    const tokenArray = new TokenArray(params.toString());
    const container = tokenArray.asFilterValuesFromResponse(response);

    // Assert
    expect(container).toMatchSnapshot();
  });

  it("creates filter container with removing constrain if there is no depepdent rows", () => {
    // Arrange
    const params = new URLSearchParams({
      "0[s0.channel]": "channel-pln",
    });
    const response = new InitialStateResponse(
      [
        {
          label: "Cat1",
          value: "cat-1-id",
          slug: "cat1",
        },
      ],
      {},
      [
        {
          label: "Channel PLN",
          value: "channel-pln",
          slug: "channel-pln",
        },
      ],
    );

    // Act
    const tokenArray = new TokenArray(params.toString());
    const container = tokenArray.asFilterValuesFromResponse(response);

    // Assert
    expect(container).toMatchSnapshot();
  });
});
