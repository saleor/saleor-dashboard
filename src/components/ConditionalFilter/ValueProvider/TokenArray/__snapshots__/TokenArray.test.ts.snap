// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConditionalFilter / ValueProvider / TokenArray creates filter container with removing constrain if there is no depepdent rows 1`] = `
TokenArray [
  FilterElement {
    "condition": Condition {
      "loading": false,
      "options": ConditionOptions [
        Object {
          "label": "is",
          "type": "select",
          "value": "input-5",
        },
      ],
      "selected": ConditionSelected {
        "conditionValue": Object {
          "label": "is",
          "type": "select",
          "value": "input-5",
        },
        "loading": false,
        "options": Array [],
        "value": Object {
          "label": "Channel PLN",
          "slug": "channel-pln",
          "value": "channel-pln",
        },
      },
    },
    "constraint": undefined,
    "loading": false,
    "value": ExpressionValue {
      "label": "Channel",
      "type": "channel",
      "value": "channel",
    },
  },
]
`;

exports[`ConditionalFilter / ValueProvider / TokenArray should create filter container from a given response 1`] = `
TokenArray [
  FilterElement {
    "condition": Condition {
      "loading": false,
      "options": ConditionOptions [
        Object {
          "label": "is",
          "type": "number",
          "value": "input-1",
        },
        Object {
          "label": "lower",
          "type": "number",
          "value": "input-2",
        },
        Object {
          "label": "greater",
          "type": "number",
          "value": "input-3",
        },
        Object {
          "label": "between",
          "type": "number.range",
          "value": "input-4",
        },
      ],
      "selected": ConditionSelected {
        "conditionValue": Object {
          "label": "is",
          "type": "number",
          "value": "input-1",
        },
        "loading": false,
        "options": Array [],
        "value": "123",
      },
    },
    "constraint": undefined,
    "loading": false,
    "value": ExpressionValue {
      "label": "Price",
      "type": "price",
      "value": "price",
    },
  },
  "AND",
  FilterElement {
    "condition": Condition {
      "loading": false,
      "options": ConditionOptions [
        Object {
          "label": "is",
          "type": "select",
          "value": "input-5",
        },
      ],
      "selected": ConditionSelected {
        "conditionValue": Object {
          "label": "is",
          "type": "select",
          "value": "input-5",
        },
        "loading": false,
        "options": Array [],
        "value": Object {
          "label": "Channel PLN",
          "slug": "channel-pln",
          "value": "channel-pln",
        },
      },
    },
    "constraint": Constraint {
      "dependsOn": Array [
        "price",
        "isVisibleInListing",
        "isAvailable",
        "isPublished",
      ],
      "disabled": Array [
        "left",
        "condition",
      ],
      "removable": false,
    },
    "loading": false,
    "value": ExpressionValue {
      "label": "Channel",
      "type": "channel",
      "value": "channel",
    },
  },
]
`;
